### PR TITLE
Intel FSP support

### PR DESCRIPTION
**Pull request by rizlik**
Initial support for Intel Architecture using the Intel Firmware Support package.

### CODE REVIEW

File: hal/x86_fsp_qemu.c [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+#define PCI_AHCI_BUS 0
+#define PCI_AHCI_DEV 31
+#define PCI_AHCI_FUN 2
+
+static uint32_t sata_bar = 0;
+
+static void init_sata(void)
+{
+    uint32_t version;
+    char rbuf[32];
```
Comment (dgarske) [inline hal/x86_fsp_qemu.c:46]: Is `rbuf` used?

File: hal/x86_fsp_tgl.c [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+#define PCI_AHCI_BUS 0
+#define PCI_AHCI_DEV 0x17
+#define PCI_AHCI_FUN 0
+
+void x86_fsp_tgl_init_sata(void)
+{
+    uint32_t sata_bar;
+    uint32_t version;
+    char rbuf[32];
```
Comment (dgarske) [inline hal/x86_fsp_tgl.c:40]: Same `rbuf` used?

File: include/user_settings.h [hunk 3]
```diff
@@ -284,4 +289,15 @@ extern int tolower(int c);
 #   define WOLFSSL_SMALL_STACK
 #endif
 
+
+#ifdef WOLFTPM_MMIO
+    void delay(int msec);
+    #define XTPM_WAIT() delay(1000);
+    #define DEBUG_WOLFTPM
+    #define WOLFTPM_ADV_IO
+    #define WOLFTPM_DEBUG_VERBOSE
```
Comment (dgarske) [inline include/user_settings.h:298]: Don't recommend verbose on by default.

File: src/x86/hob.c [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+                        hob->u.resource_descriptor.resource_type);
+        dump_guid(&hob->u.resource_descriptor.owner);
+        wolfBoot_printf("Physical Start: 0x%x_%x\r\n",
+                        (uint32_t)(hob->u.resource_descriptor.physical_start >> 32),
+                        (uint32_t)hob->u.resource_descriptor.physical_start);
+        wolfBoot_printf("Resource Length: 0x%x_%x\r\n",
+                        (uint32_t)(hob->u.resource_descriptor.resource_length >> 32),
+                        (uint32_t)hob->u.resource_descriptor.resource_length
+                        );
+        // Handle specific fields for this HOB type
```
Comment (dgarske) [inline src/x86/hob.c:198]: Use C style `/* */` comments.

File: docs/Targets.md [hunk 5]
```diff
@@ -1572,3 +1572,87 @@ Flash Allocation:
 ```
 
 Detailed steps can be found at [Readme](../IDE/Renesas/e2studio/RA6M4/Readme.md).
+
+
+## Intel x86_64 with Intel FSP support
```
Comment (dgarske) [inline docs/Targets.md:1577]: Can you include sample boot output?

**Review by dgarske — CHANGES_REQUESTED** [review]
Amazing work on this!

File: src/x86/common.c [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+             "mov $0xc0000080, %%ecx\r\n"
+             "rdmsr\r\n"
+             "or %1, %%eax\r\n"
+             "mov $0xc0000080, %%ecx\r\n"
+             "wrmsr\r\n"
+             "mov %%cr0, %%eax\r\n"
+             "or %2, %%eax\r\n"
+             "mov %%eax, %%cr0\r\n"
+             "mov %3, %%eax\r\n"
+             "ljmp %5,$_setcs\r\n"
```
Comment (dgarske) [inline src/x86/common.c:221]:
```	[CC-x86_64] common.o
../src/x86/common.c: Assembler messages:
../src/x86/common.c:220: Error: can't handle non absolute segment in `ljmp'
make[1]: *** [Makefile:160: common.o] Error 1```

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/x86/hob.c (hunk 4 above)
Comment (rizlik) [reply src/x86/hob.c]: the code was unused, it has been removed

File: docs/Targets.md (hunk 5 above)
Comment (rizlik) [reply docs/Targets.md:1577]: working on that: other parts of the documentation need to be updated

**Review by dgarske — CHANGES_REQUESTED** [review]
Still getting:

```
	[CC-x86_64] common.o
gcc -D"__WOLFBOOT" -Werror -Wextra -DARCH_x86_64 -DFORCE_32BIT -DWOLFSSL_SP_DIV_WORD_HALF -DWOLFBOOT_FSP=1 -DWOLFBOOT_TGL=1 -fno-stack-protector -m32 -fno-PIC -fno-pie -mno-mmx -mno-sse -DDEBUG_UART -DUCODE0_ADDRESS=0xfffc4000 -DARCH_FLASH_OFFSET=0x0 -DWOLFBOOT_ORIGIN=0xffff0000 -DBOOTLOADER_PARTITION_SIZE=0x400000 -DWOLFBOOT_ARCH=x86_64 -DTARGET_kontron_vx3060_s2 -D"WOLFBOOT_SIGN_ECC256" -D"FILL_BYTE=0xFF" -O0 -D"DEBUG" -g -ggdb3 -D"WOLFBOOT_HASH_SHA256" -DIMAGE_HEADER_SIZE=256 -DWOLFBOOT_64BIT -DX86_UART_BASE=0xFE032000 -DX86_UART_REG_WIDTH=4 -DX86_UART_MMIO=1 -DPCH_HAS_PCR=1 -DPCI_USE_ECAM=1 -DPCH_PCR_BASE=0xFD000000 -DPCI_ECAM_BASE=0xC0000000 -DFSP_S_UPD_DATA_BASE=0xffe35000 -DWOLFBOOT_LOAD_BASE=0x2000000  -I".." -I"../include/" -I"../lib/wolfssl" -D"WOLFSSL_USER_SETTINGS" -D"WOLFTPM_USER_SETTINGS" -D"PLATFORM_kontron_vx3060_s2" -DBUILD_LOADER_STAGE1 -DWOLFBOOT_STAGE1_SIZE=0x1000 -DWOLFBOOT_STAGE1_LOAD_ADDR= -DWOLFBOOT_STAGE1_FLASH_ADDR= -DWOLFBOOT_STAGE1_BASE_ADDR= -c -o common.o ../src/x86/common.c
../src/x86/common.c: Assembler messages:
../src/x86/common.c:220: Error: can't handle non absolute segment in `ljmp'
make[1]: *** [Makefile:160: common.o] Error 1
```

```
davidgarske@P700-XEON:~/GitHub/wolfboot$ which gcc
/usr/bin/gcc
davidgarske@P700-XEON:~/GitHub/wolfboot$ gcc --version
gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

File: hal/x86_fsp_tgl.c (hunk 2 above)
Comment (danielinux) [reply hal/x86_fsp_tgl.c]: removed in b1c1ba4 - thank you

**Review by dgarske — APPROVED** [review]

**Comment by dgarske** [comment]
@rizlik and @danielinux let me know when you are ready for merge. Shall I squash it?

**Review by dgarske — APPROVED** [review]

**Pull request merged**